File: entities/audit.go
package entities

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type AuditEvent struct {
	EventID      uuid.UUID `gorm:"primaryKey" json:"event_id"`
	ActorType    string    `gorm:"size:50;not null;index" json:"actor_type"`
	ActorID      uuid.UUID `gorm:"not null;index" json:"actor_id"`
	Action       string    `gorm:"size:100;not null;index" json:"action"`
	ResourceType string    `gorm:"size:100;not null" json:"resource_type"`
	ResourceID   string    `gorm:"size:255;not null" json:"resource_id"`
	BeforeState  *string   `gorm:"type:text" json:"before_state,omitempty"`
	AfterState   *string   `gorm:"type:text" json:"after_state,omitempty"`
	IPAddress    *string   `gorm:"size:50" json:"ip_address,omitempty"`
	Device       *string   `gorm:"type:text" json:"device,omitempty"`
	Timestamp    time.Time `gorm:"autoCreateTime" json:"timestamp"`
}

func (a *AuditEvent) BeforeCreate(tx *gorm.DB) (err error) {
	if a.EventID == uuid.Nil {
		a.EventID = uuid.New()
	}
	return nil
}

File: backend_api/entities/audit.py
import uuid
from sqlalchemy import Column, String, DateTime, Text
from sqlalchemy.sql import func
from database import Base

class AuditEvent(Base):
    __tablename__ = "audit_events"

    event_id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    actor_type = Column(String(50), nullable=False, index=True)
    actor_id = Column(String(36), nullable=False, index=True)
    action = Column(String(100), nullable=False, index=True)
    resource_type = Column(String(100), nullable=False)
    resource_id = Column(String(255), nullable=False)
    before_state = Column(Text, nullable=True)
    after_state = Column(Text, nullable=True)
    ip_address = Column(String(50), nullable=True)
    device = Column(Text, nullable=True)
    timestamp = Column(DateTime(timezone=True), server_default=func.now())
